src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,511 @@
+}
+
+/* ----------------------------------------------------------------------
+   parse optional parameters
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::options(int narg, char **arg)
+{
+    if (narg < 0) error->all(FLERR,"Illegal dynamical_matrix command");
+    int iarg = 0;
+    const char* filename = "dynmat.dyn";
+    while (iarg < narg) {
+        if (strcmp(arg[iarg],"binary") == 0) {
+            if (iarg + 2 > narg) error->all(FLERR, "Illegal dynamical_matrix command");
+            if (strcmp(arg[iarg+1],"gzip") == 0) {
+                compressed = 1;
+            }
+            else if (strcmp(arg[iarg+1],"yes") == 0) {
+                binaryflag = 1;
+            }
+            iarg += 2;
+        }
+        else if (strcmp(arg[iarg],"file") == 0) {
+            if (iarg+2 > narg) error->all(FLERR, "Illegal dynamical_matrix command");
+            filename = arg[iarg + 1];
+            file_flag = 1;
+            iarg += 2;
+        } else error->all(FLERR,"Illegal dynamical_matrix command");
+    }
+    if (file_flag == 1) {
+        openfile(filename);
+    }
+}
+
+/* ----------------------------------------------------------------------
+   generic opening of a file
+   ASCII or binary or gzipped
+   some derived classes override this function
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::openfile(const char* filename)
+{
+    // if file already opened, return
+    if (me!=0) return;
+    if (file_opened) return;
+
+    if (compressed) {
+#ifdef LAMMPS_GZIP
+        char gzip[128];
+    sprintf(gzip,"gzip -6 > %s",filename);
+#ifdef _WIN32
+    fp = _popen(gzip,"wb");
+#else
+    fp = popen(gzip,"w");
+#endif
+#else
+        error->one(FLERR,"Cannot open gzipped file");
+#endif
+    } else if (binaryflag) {
+        fp = fopen(filename,"wb");
+    } else {
+        fp = fopen(filename,"w");
+    }
+
+    if (fp == NULL) error->one(FLERR,"Cannot open dump file");
+
+    file_opened = 1;
+}
+
+/* ----------------------------------------------------------------------
+   create dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::calculateMatrix(char *arg)
+{
+    int nlocal = atom->nlocal;
+    int plocal;
+    int id;
+    int group_flag = 0;
+    int *mask = atom->mask;
+    int *type = atom->type;
+    tagint *aid = atom->tag; //atom id
+    double dyn_element[nlocal][3]; //first index is nlocal (important)
+    double moaoi; //mass of atom of interest
+    double imass;
+    double del = force->numeric(FLERR, arg);
+    double *m = atom->mass;
+    double **x = atom->x;
+    double **f = atom->f;
+
+
+    //initialize dynmat to all zeros
+    for (int i=0; i < dynlen; i++)
+        for (int j=0; j < dynlen; j++)
+            dynmat[i][j] = 0.;
+
+
+    if (strstr(arg,"v_") == arg) error->all(FLERR,"Variable for dynamical_matrix is not supported");
```

Comment:
Same as above, this should be tested outside, and it will never catch, anyway, since `force->numeric()` will refuse a non-numeric string `arg`, anyway.